medschool/src/parse.rs

```diff
@@ -0,0 +1,286 @@
+
+use std::collections::{BTreeSet, HashMap, HashSet};
+
+use syn::{Attribute, Expr, Ident, Meta, Type, TypePath};
+
+use crate::{
+    types::{PartialField, PartialType},
+    DocsError,
+};
+
+#[tracing::instrument(level = "trace", ret)]
+pub fn docs_from_attributes(attributes: Vec<Attribute>) -> Vec<String> {
+    attributes
+        .into_iter()
+        .filter_map(|attribute| {
+            if let Meta::NameValue(meta_doc) = attribute.meta {
+                if let (ident, Expr::Lit(lit)) = (
+                    meta_doc.path.segments.first()?.ident.clone(),
+                    meta_doc.value,
+                ) {
+                    if ident == Ident::new("doc", ident.span()) {
+                        if let syn::Lit::Str(lit_str) = lit.lit {
+                            let doc = lit_str.value();
+                            return Some(if doc.trim().is_empty() {
+                                "\n".to_string()
+                            } else {
+                                format!("{}\n", doc)
+                            });
+                        }
+                    }
+                }
+            }
+            None
+        })
+        .collect()
+}
+
+/// Digs into the generics of a field ([`syn::ItemStruct`] only), trying to get the last
+/// [`syn::PathArguments`], which (hopefully) contains the concrete type we care about.
+///
+/// Extracts `Foo` from `foo: Option<Vec<{Foo}>>`.
+///
+/// Doesn't handle generics of generics though, so if your field is `baz: Option<T>` we're going
+/// to be assigning this field type to be the string `"T"` (which is probably not what you
+/// wanted).
+#[tracing::instrument(level = "trace", ret)]
+pub fn get_ident_from_field_skipping_generics(type_path: TypePath) -> Option<Ident> {
+    // welp, this path probably contains generics
+    let segment = type_path.path.segments.last()?;
+
+    let mut current_argument = segment.arguments.clone();
+    let mut inner_type = None;
+
+    // keep digging into the `PathArguments`
+    while let syn::PathArguments::AngleBracketed(generics) = &current_argument {
+        // go directly to the last piece of generics, skipping lifetimes
+        match generics.args.last()? {
+            // finally have something that resembles a type, but might be an `Option`, so
+            // we have to go deeper!
+            syn::GenericArgument::Type(Type::Path(generic_path)) => {
+                // that's it, we've reached the final type
+                inner_type = match generic_path.path.segments.last() {
+                    Some(t) => Some(t.ident.clone()),
+                    None => break,
+                };
+
+                current_argument = generic_path.path.segments.last()?.arguments.clone();
+            }
+            _ => break,
+        }
+    }
+
+    inner_type
+}
+
+/// Converts a [`syn::ItemMod`] into a [`PartialType`].
+fn parse_item_mod(item_mod: syn::ItemMod) -> Option<PartialType> {
```

Comment:
Why is this an `Option` if the return is always `Some`?

Reply:
because other related functions do, like `parse_item_struct` and can return None. all these functions are used in `parse_syn_item_into_partial_type`